src/main/java/htsjdk/samtools/SamFiles.java

```diff
@@ -19,61 +21,73 @@
      * @return The index for the provided SAM, or null if one was not found.
      */
     public static File findIndex(final File samFile) {
-        final File indexFile = lookForIndex(samFile); //try to find the index
-        if (indexFile == null) {
-            return unsymlinkAndLookForIndex(samFile);
+        Path path = findIndex(samFile.toPath());
+        return path == null ? null : path.toFile();
+    }
+
+    /**
+     * Finds the index file associated with the provided SAM file.  The index file must exist and be reachable to be found.
+     *
+     * If the file is a symlink and the index cannot be found, try to unsymlink the file and look for the bai in the actual file path.
+     *
+     * @return The index for the provided SAM, or null if one was not found.
+     */
+    public static Path findIndex(final Path samPath) {
+        final Path indexPath = lookForIndex(samPath); //try to find the index
+        if (indexPath == null) {
+            return unsymlinkAndLookForIndex(samPath);
         } else {
-            return indexFile;
+            return indexPath;
         }
     }
 
     /**
      * resolve the canonical path of samFile and attempt to find an index there.
      * @return an index file or null if no index is found.
      */
-    private static File unsymlinkAndLookForIndex(File samFile) {
+    private static Path unsymlinkAndLookForIndex(Path samPath) {
         try {
-            final File canonicalSamFile = samFile.getCanonicalFile();
-            final File canonicalIndexFile = lookForIndex(canonicalSamFile);
-            if ( canonicalIndexFile != null) {
-                System.err.println("The index file " + canonicalIndexFile.getPath()
+            final Path canonicalSamPath = samPath.toRealPath();
+            final Path canonicalIndexPath = lookForIndex(canonicalSamPath);
+            if ( canonicalIndexPath != null) {
+                System.err.println("The index file " + canonicalIndexPath.toAbsolutePath()
                         + " was found by resolving the canonical path of a symlink: "
-                        + samFile.getPath() + " -> " + samFile.getCanonicalPath());
+                        + samPath.toAbsolutePath() + " -> " + samPath.toRealPath());
             }
-            return canonicalIndexFile;
+            return canonicalIndexPath;
         } catch (IOException e) {
             return null;
         }
     }
 
-    private static File lookForIndex(final File samFile) {// If input is foo.bam, look for foo.bai
-        File indexFile;
-        final String fileName = samFile.getName();
+    private static Path lookForIndex(final Path samPath) {// If input is foo.bam, look for foo.bai
+        Path indexPath;
+        final String fileName = samPath.getFileName().toString();
```

Comment:
Yes, it will. `samPath.getFileName()` returns the last component of the Path, which is then converted to a string.